---
title: "Ataxia RNAseq - STAR junction category distributions"
author: 
- name: "Guillermo Rocamora PÃ©rez"
  affiliation: UCL - Version 1.0
date: "`r format(Sys.time(), '%d %B %Y')`"
output:
  bookdown::html_document2:
    figure_caption: yes
    code_folding: show
    theme: paper
    highlight: haddock
    df_print: paged
    toc: true
    toc depth: 3
    toc_float: true
    number_sections: true
  md_document:
    variant: markdown_github
    toc: true
    number_sections: true
always_allow_html: true
---

```{r setup, include = F}
## Load libraries
shhh <- suppressPackageStartupMessages
shhh(library(here))
shhh(library(SummarizedExperiment))
shhh(library(ggsci))
shhh(library(tidyverse))
shhh(library(foreach))
shhh(library(doParallel))

main_path <- file.path("/home/grocamora/Core_Projects/Ataxia_Splicing_Analysis/")

## Load helper functions
source(file.path(main_path, "R/STAR_introns.R"))
source(file.path(main_path, "R/hf_graph_and_themes.R"))
source(file.path(main_path, "R/hf_additional.R"))

## Set options
options(dplyr.summarise.inform = FALSE)
options(lifecycle_verbosity = "warning")
knitr::opts_chunk$set(echo = F, warning = F, message = F, out.width="100%", fig.align = "center", dpi = 200)
```

```{r load-theme, echo = F, results = "asis"}
## Custom sciRmdTheme with increased width.
## Please ask guillermorocamora@gmail.com for more details.
# if(knit_output == "html"){
#   sciRmdTheme::set.theme(
#     theme = "default",
#     color = NULL,
#     header.sticky = FALSE,
#     list.group.icon = "arrow",
#     font.family = "Arial",
#     font.color = "black",
#     header.color = "darkblue"
#   )
# }

## Custom styles for the output html
cat('
<style type="text/css">
.dataTables_scrollHeadInner{
  width:100% !important;
}
.dataTables_scrollHeadInner table{
  width:100% !important;
}
/*
.code-folding-btn {
  display: none;
}
*/
h3, .h3 {
  font-size: 22px!important;
}
h4, .h4 {
  font-size: 18px!important;
}
h5, .h5 {
  font-size: 16px!important;
}
body{
  font-size: 13px;
}
.tocify-subheader {
    text-indent: 15px;
    display: none;
    font-size: 12px;
}
</style>')
```

# Background
> Aim: update the studies carried out by Regina H. Reynolds about STAR introns' annotation distributions. Source material: [STAR: intron annotation](https://rhreynolds.github.io/LBD-seq-bulk-analyses/quantification_splicing/STAR_intron_annotation.html)

This report is meant to be used as a template. As such, the different sections will include instructions on how to modify the graphic outputs and not descriptions about the results themselves.

Three different studies will be presented: disease status (Level 1), ataxia subtypes (Level 2) and diagnoses (Level 3). To run this script, it is necessary to source the file at `/home/grocamora/Core_Projects/Ataxia_Splicing_Analysis/R/STAR_introns.R` with needed helper functions.

```{r, echo = T}
# Paths
metadata_path <- file.path(main_path, "metadata/metadata.csv")
multiqc_path <- file.path(main_path, "metadata/multiqc_rseqc_read_distribution.txt")
sj_tab_path <- file.path(main_path, "samples/SJ_out_tab_files/")

# Path to the gtf file to annotate the introns and the ENCODE blacklisted region
#gtf_ensembl_path <- file.path("/home/grocamora/RytenLab-Research/Additional_files/Homo_sapiens.GRCh38.105.gtf")
gtf_gencode_path <- file.path("/home/grocamora/RytenLab-Research/Additional_files/GENCODE/gencode.v43.annotation.gtf")
blacklist_path = file.path("/home/grocamora/RytenLab-Research/Additional_files/hg38-blacklist.v2.bed")

# Load the metadata
metadata <- readr::read_delim(metadata_path, show_col_types = FALSE) %>% 
  dplyr::rowwise() %>%
  dplyr::mutate(Individual_ID = stringr::str_split(ID_anon, "_", simplify = T)[1]) %>%
  dplyr::filter(!(Diagnosis %in% c("CANVAS", "AIFM1"))) %>%
  extractReadDepthMultiQC(multiqc_path)
```

# Pipeline

## Load the SJ files

The first step is to load the junctions from the STAR alignment output files (`SJ.tab.out`). We employ the [`loadSJ`](https://github.com/guillermo1996/Ataxia_Splicing_Analysis/blob/main/R/STAR_introns.R#L35) function, an adaptation of the `load_sj_df` function by Regina. [Source](https://github.com/RHReynolds/LBD-seq-bulk-analyses/blob/main/R/load_sj_df.R).

```{r, echo = T}
sj_df <- loadSJ(metadata, sj_tab_path)
```

Following is an example of the dataframe generated (minus some columns not employed in the analysis):

```{r}
sj_df %>%
  dplyr::sample_n(10) %>%
  dplyr::select(-c(intron_motif, intron_annotation, multimap_reads_junction, max_splice_alignment_overhang)) %>%
  kableExtra::kbl(booktabs = T, linesep = "") %>%
  kableExtra::kable_classic(full_width = F, "hover", "striped", html_font = "Cambria", font_size = 12) %>%
  kableExtra::row_spec(0, bold = T, font_size = 14)
```

## Create the RSE_jx object

Next, we create the RSE object with the annotated junctions, the reads matrix and the sample metadata. We employ the function [`createRSE`](https://github.com/guillermo1996/Ataxia_Splicing_Analysis/blob/main/R/STAR_introns.R#L90), upgraded and adapted from the function `create_rse_jx` by Regina. [Source](https://github.com/RHReynolds/LBD-seq-bulk-analyses/blob/main/R/create_rse_jx.R).

The main difference is that only junctions from chromosomes 1:22, X and Y are considered (can be modified in the function with the variable `valid_seqnames`).

Given that the annotation process is long, we only create the variable if it does not exists in disk.

```{r, echo = T}
rse_jx_path <- file.path(main_path, "variables/rse_jx_gencode43.rds")
if(!file.exists(rse_jx_path)){
  rse_jx <- createRSE(sj_df, gtf_gencode_path, sample_info = metadata, sample_id_col = "Correct_sample")
  rse_jx %>% saveRDS(rse_jx_path)
}else{
  rse_jx <- readRDS(rse_jx_path)
}
```

## Removing background noise

This involves:

* Removing all junctions that overlap with the ENCODE blacklisted regions.

* Removing all junctions that are associated to multiple genes (this removes all junctions from the category `ambig. gene` and some of the `annotated`).

* Removing all junctions that are less than 25 bp long.

```{r, echo = T}
encode_blacklist_hg38 <- loadEncodeBlacklist(blacklist_path)
rse_jx <- removeEncodeBlacklistRegionsRSE(rse_jx, encode_blacklist_hg38)
rse_jx <- removeAmbiguousGenesRSE(rse_jx)
rse_jx <- removeShortJunctionsRSE(rse_jx)
```

# Visualization of all junction categories

The first plots shown here represent the distributions using all categories from the `dasper` annotation function. As mentioned, results are split in the different levels of study.

Across all visualizations, a Wilcoxon signed-rank test between the distributions were executed, and significant results ($p<0.05$) are shown in the graphs. The p-values are Bonferroni corrected given the number of tests executed per group.

```{r}
category_names <- c("annotated" = "Annotated",
                    "novel_acceptor" = "Novel Acceptor",  
                    "novel_donor" = "Novel Donor", 
                    "diff_acceptor_donor" = "Diff. Acceptor/Donor",
                    "novel_exon_skip" = "Novel Exon Skip", 
                    "novel_combo" = "Novel Combo", 
                    "ambig_gene" = "Ambig. genes", 
                    "unannotated" = "Unannotated")

# Get all reads per junction and sample
annot_junc_df <- rowRanges(rse_jx)[, "type"] %>%
  S4Vectors::mcols() %>%
  tibble::as_tibble(rownames = "junction_id") %>%
  dplyr::mutate(type = factor(category_names[type %>% as.character()], levels = category_names)) %>%
  dplyr::left_join(assay(rse_jx) %>% as_tibble(rownames = "junction_id"), by = "junction_id") %>%
  tidyr::pivot_longer(cols = -c("junction_id", "type"), names_to = "sample_id", values_to = "count") %>%
  dplyr::filter(count > 0)

# Summarize the proportions by junction category and sample
annot_junc_prop_df <- annot_junc_df %>% 
  dplyr::group_by(sample_id, type) %>%
  dplyr::summarise(n_junc = n(), count = sum(count)) %>%
  dplyr::left_join(colData(rse_jx) %>% 
                     tibble::as_tibble(rownames = "sample_id") %>%
                     dplyr::select(sample_id, total_reads),
                   by = "sample_id") %>%
  dplyr::group_by(sample_id) %>%
  dplyr::mutate(total_junc = sum(n_junc),
                total_count = sum(count),
                prop_junc = n_junc/total_junc,
                prop_count = count/total_count,
                prop_total_reads = count/total_reads) %>%
  dplyr::select(-total_reads) %>%
  dplyr::ungroup() %>%
  # tidyr::pivot_longer(names_to = "metric", values_to = "value", cols = -c(type, sample_id)) %>%
  # tidyr::pivot_wider(names_from = "type", values_from = "value") %>%
  # dplyr::mutate(`Diff. Acceptor/Donor` = `Novel Acceptor` - `Novel Donor`) %>%
  # tidyr::pivot_longer(names_to = "type", values_to = "value", -c(sample_id, metric)) %>%
  # tidyr::pivot_wider(names_from = "metric", values_from = "value") %>%
  # dplyr::mutate(type = factor(type, levels = category_names)) %>%
  dplyr::left_join(colData(rse_jx) %>% 
                     tibble::as_tibble(rownames = "sample_id") %>%
                     dplyr::select(sample_id, Type, Diagnosis, AtaxiaSubtype, Region),
                   by = "sample_id") %>%
  dplyr::mutate(Type = factor(Type, levels = c("Control", "Case")))
```

## Proportion of junctions by category {.tabset}

First, we plot the category percentages by sample. We employ the function `plotJunctionCategories`, which allows to select a different level to group by (three different possibilities: `Type`, `AtaxiaSubtype` and `Diagnosis`). [Source](https://rhreynolds.github.io/LBD-seq-bulk-analyses/quantification_splicing/STAR_intron_annotation.html#4_Proportion_annotatedpartially_annotatedunannotated_per_group).

```{r, echo = T}
#' Plots the category percentages by sample
#'
#' @param annot_junc_prop_df dataframe, contains the proportion of junctions for
#'   each category and each sample.
#' @param level character vector, specifies the level of study. Name of the
#'   column that will split the graph and the Wilcoxon tests. For the Ataxia
#'   RNAseq data, only "Type", "AtaxiaSubtype" and "Diagnosis" are valid inputs.
#' @param split_tissue boolean, whether to split the graph in facets by tissue.
#' @param ref_group character vector, category in the \code{level} field that
#'   will be compare against for the Wilcoxon signed-rank test.
#'
#' @return
#' @export
plotJunctionCategories <- function(annot_junc_prop_df, level, split_tissue = F, ref_group = NULL){
  p <- ggplot(annot_junc_prop_df,
              aes(x = type, y = prop_junc, fill = !!sym(level))) +
    geom_boxplot() +
    scale_y_continuous(expand = expansion(mult = c(0.01, 0.15))) +
    labs(x = "", y = "Proportion of junctions") +
    scale_fill_manual(values = pal_jco("default", alpha = 0.9)(10)[c(1, 9)]) +
    ggpubr::geom_pwc(aes(group = !!sym(level)), label = " p = {p.adj.format}", p.adjust.method = "bonferroni", 
                     p.adjust.by = "group", vjust = -0.5, hide.ns = T, ref.group = ref_group) +
    custom_gg_theme
  if(split_tissue){
    p <- p + 
      facet_wrap(vars(Region), ncol = 1) + 
      scale_fill_manual(values = pal_jco("default", alpha = 0.9)(10)[c(3, 1, 9, 2, 7)])
  }
  
  return(p)
}
```

### Level 1 (Type) {-}

```{r, fig.height=5, fig.width=7.2, fig.cap="Proportion of junctions by disease status and brain tissue."}
plotJunctionCategories(annot_junc_prop_df, level = "Type", split_tissue = T)
```

### Level 2 (AtaxiaSubtype) {-}

```{r, fig.height=5, fig.width=7.2, fig.cap="Proportion of junctions by Ataxia Subtype and tissue."}
plotJunctionCategories(annot_junc_prop_df, level = "AtaxiaSubtype", split_tissue = T)
```

### Level 3 (Diagnosis) {-}

```{r, fig.height=5, fig.width=7.2, fig.cap="Proportion of junctions by Diagnosis and tissue."}
plotJunctionCategories(annot_junc_prop_df %>%
                         dplyr::filter(Diagnosis %in% c("Control", "FRDA", "SCA1", "SCA2", "SCA6")), 
                       level = "Diagnosis", split_tissue = T)
```

## Proportion of counts by category {.tabset}

Next, we also represent the proportion of counts (or reads) that correspond to each category when compared against the total number of junction reads. We use the function `plotJunctionCountCategories`.

```{r, echo = T}
#' Plots the proportion of counts for each category and sample
#'
#' @param annot_junc_prop_df dataframe, contains the proportion of counts for
#'   each category and sample.
#' @param level character vector, specifies the level of study. Name of the
#'   column that will split the graph and the Wilcoxon tests. For the Ataxia
#'   RNAseq data, only "Type", "AtaxiaSubtype" and "Diagnosis" are valid inputs.
#' @param tissue character vector, tissue to use to graph. Leave as NULL if no
#'   distinction is used.
#' @param ref_group character vector, category in the \code{level} field that
#'   will be compare against for the Wilcoxon signed-rank test.
#'
#' @return
#' @export
plotJunctionCountCategories <- function(annot_junc_prop_df, level, tissue = NULL, ref_group = NULL){
  # If tissue is provided, we filter the dataframe.
  if(!is.null(tissue)) annot_junc_prop_df <- annot_junc_prop_df %>% dplyr::filter(Region == tissue)
  
  p <- ggplot(annot_junc_prop_df, aes(x = !!sym(level), y = prop_count, fill = !!sym(level))) +
    geom_boxplot() +
    scale_y_continuous(expand = expansion(mult = c(0.1, 0.2))) +
    labs(x = "", y = "Proportion of junctions") +
    scale_fill_manual(values = pal_jco("default", alpha = 0.9)(10)[c(1, 9)]) +
    facet_wrap(vars(type), ncol = 3, scales = "free") +
    ggpubr::geom_pwc(aes(group = !!sym(level)), ref.group = ref_group, vjust = -0.3, hide.ns = T,
                     label = " p = {p.adj.format}", p.adjust.method = "bonferroni", p.adjust.by = "group") +
    custom_gg_theme_subtitle# + theme(axis.text.x = element_blank(), axis.ticks.length.x = unit(0, "cm"))
  
  if(!is.null(tissue)){
    p <- p +
      scale_fill_manual(values = pal_jco("default", alpha = 0.9)(10)[c(3, 1, 9, 2, 7)]) + 
      ggtitle("", subtitle = paste0("Tissue: ", tissue))
  }
  
  return(p)
}
```


### Level 1 (Type) {.tabset -}

#### Cerebellum {-}

```{r, fig.height=5, fig.width=9, fig.cap="Proportion of counts by disease status and brain tissue."}
plotJunctionCountCategories(annot_junc_prop_df, level = "Type", tissue = "Cerebellum")
```

#### Frontal {-}

```{r, fig.height=5, fig.width=9, fig.cap="Proportion of counts by disease status and brain tissue."}
plotJunctionCountCategories(annot_junc_prop_df, level = "Type", tissue = "Frontal")
```

### Level 2 (AtaxiaSubtype) {.tabset -}

#### Cerebellum {-}

```{r, fig.height=5, fig.width=9, fig.cap="Proportion of counts by Ataxia Subtype for Cerebellum tissue."}
plotJunctionCountCategories(annot_junc_prop_df, level = "AtaxiaSubtype", tissue = "Cerebellum")
```

#### Frontal Cortex {-}

```{r, fig.height=5, fig.width=9, fig.cap="Proportion of counts by Ataxia Subtype for Frontal Cortex tissue."}
plotJunctionCountCategories(annot_junc_prop_df, level = "AtaxiaSubtype", tissue = "Frontal")
```

### Level 3 (Diagnosis) {.tabset -}

#### Cerebellum {-}

```{r, fig.height=5, fig.width=9, fig.cap="Proportion of counts by Diagnosis for Cerebellum tissue."}
plotJunctionCountCategories(annot_junc_prop_df %>%
                              dplyr::filter(Diagnosis %in% c("Control", "FRDA", "SCA1", "SCA2", "SCA6")), 
                            level = "Diagnosis", tissue = "Cerebellum", ref_group = "Control")
```

#### Frontal Cortex {-}

```{r, fig.height=5, fig.width=9, fig.cap="Proportion of counts by Diagnosis for Frontal Cortex tissue."}
plotJunctionCountCategories(annot_junc_prop_df %>%
                              dplyr::filter(Diagnosis %in% c("Control", "FRDA", "SCA1", "SCA2", "SCA6")), 
                            level = "Diagnosis", tissue = "Frontal", ref_group = "Control")
```

# Visualization for three categories

Next, the same visualizations are presented but using only three categories: `Annotated`, `Partially annotated` and `Unannotated`.

```{r}
category_names_red <- c("annotated" = "Annotated",
                        "novel_acceptor" = "Partially Annotated",  
                        "novel_donor" = "Partially Annotated", 
                        "novel_exon_skip" = "Partially Annotated", 
                        "novel_combo" = "Partially Annotated", 
                        "ambig_gene" = "Ambig. genes", 
                        "unannotated" = "Unannotated")

# Get all reads per junction and sample
annot_junc_red_df <- rowRanges(rse_jx)[, "type"] %>%
  S4Vectors::mcols() %>%
  tibble::as_tibble(rownames = "junction_id") %>%
  dplyr::mutate(type = factor(category_names_red[type %>% as.character()], levels = unique(category_names_red))) %>%
  dplyr::left_join(assay(rse_jx) %>% as_tibble(rownames = "junction_id"), by = "junction_id") %>%
  tidyr::pivot_longer(cols = -c("junction_id", "type"), names_to = "sample_id", values_to = "count") %>%
  dplyr::filter(count > 0)

# Summarize the proportions by junction category and sample
annot_junc_prop_red_df <- annot_junc_red_df %>% 
  dplyr::group_by(sample_id, type) %>%
  dplyr::summarise(n_junc = n(), count = sum(count)) %>%
  dplyr::left_join(colData(rse_jx) %>% 
                     tibble::as_tibble(rownames = "sample_id") %>%
                     dplyr::select(sample_id, total_reads),
                   by = "sample_id") %>%
  dplyr::group_by(sample_id) %>%
  dplyr::mutate(total_junc = sum(n_junc),
                total_count = sum(count),
                prop_junc = n_junc/total_junc,
                prop_count = count/total_count,
                prop_total_reads = count/total_reads) %>%
  dplyr::select(-total_reads) %>%
  dplyr::left_join(colData(rse_jx) %>% 
                     tibble::as_tibble(rownames = "sample_id") %>%
                     dplyr::select(sample_id, Type, Diagnosis, AtaxiaSubtype, Region),
                   by = "sample_id")
```

## Proportion of junctions by category {.tabset}

### Level 1 (Type) {-}

```{r, fig.height=5, fig.width=7.2, fig.cap="Proportion of junctions by disease status and brain tissue."}
plotJunctionCategories(annot_junc_prop_df, level = "Type", split_tissue = T)
```

### Level 2 (AtaxiaSubtype) {-}

```{r, fig.height=5, fig.width=7.2, fig.cap="Proportion of junctions by Ataxia Subtype and tissue."}
plotJunctionCategories(annot_junc_prop_red_df, level = "AtaxiaSubtype", split_tissue = T)
```

### Level 3 (Diagnosis) {-}

```{r, fig.height=5, fig.width=7.2, fig.cap="Proportion of junctions by Diagnosis and tissue."}
plotJunctionCategories(annot_junc_prop_red_df %>%
                         dplyr::filter(Diagnosis %in% c("Control", "FRDA", "SCA1", "SCA2", "SCA6")), 
                       level = "Diagnosis", split_tissue = T)
```

## Proportion of counts by category {.tabset}

### Level 1 (Type) {.tabset -}

#### Cerebellum {-}

```{r, fig.height=3, fig.width=9, fig.cap="Proportion of counts by disease status and brain tissue."}
plotJunctionCountCategories(annot_junc_prop_red_df, level = "Type", tissue = "Cerebellum")
```

#### Frontal {-}

```{r, fig.height=3, fig.width=9, fig.cap="Proportion of counts by disease status and brain tissue."}
plotJunctionCountCategories(annot_junc_prop_red_df, level = "Type", tissue = "Frontal")
```

### Level 2 (AtaxiaSubtype) {.tabset -}

#### Cerebellum {-}

```{r, fig.height=3, fig.width=9, fig.cap="Proportion of counts by Ataxia Subtype for Cerebellum tissue."}
plotJunctionCountCategories(annot_junc_prop_red_df, level = "AtaxiaSubtype", tissue = "Cerebellum")
```

#### Frontal Cortex {-}

```{r, fig.height=3, fig.width=9, fig.cap="Proportion of counts by Ataxia Subtype for Frontal Cortex tissue."}
plotJunctionCountCategories(annot_junc_prop_red_df, level = "AtaxiaSubtype", tissue = "Frontal")
```

### Level 3 (Diagnosis) {.tabset -}

#### Cerebellum {-}

```{r, fig.height=3, fig.width=9, fig.cap="Proportion of counts by Diagnosis for Cerebellum tissue."}
plotJunctionCountCategories(annot_junc_prop_red_df %>%
                              dplyr::filter(Diagnosis %in% c("Control", "FRDA", "SCA1", "SCA2", "SCA6")), 
                            level = "Diagnosis", tissue = "Cerebellum", ref_group = "Control")
```

#### Frontal Cortex {-}

```{r, fig.height=3, fig.width=9, fig.cap="Proportion of counts by Diagnosis for Frontal Cortex tissue."}
plotJunctionCountCategories(annot_junc_prop_red_df %>%
                              dplyr::filter(Diagnosis %in% c("Control", "FRDA", "SCA1", "SCA2", "SCA6")), 
                            level = "Diagnosis", tissue = "Frontal", ref_group = "Control")
```
